Add structures for 32bit paging

woppopo opened the pull request:
This is almost a copy of `src/bits64/paging.rs` (including utests).

Changes from 64bit one:
- u64 -> u32
- In 32bit, PAGE_SIZE_ENTRIES is 1024 because entry is 32bit length
- Page directory index and page table index are 10bit length (`pd_index`, `pt_index`)
- ADDRESS_MASK just turn off flag bits
- Variables and structures about page directory pointer table is removed

I'm not knowledgeable about x86 architecture, so there may be a mistake.

Question: Why does only VAddr have as_mut_ptr and as_ptr?

gz commented:
Hi thanks, I will check it soon but on a first glance this looks good to me.

The idea behind why PAddr doesn't have `as_mut_ptr` etc. is that PAddr is not supposed to be accessible (e.g., only if you would set up your pages with a 1:1 identity mapping from physical to virtual -- but then you could argue that the phys address is a virtual address..) However, of course that means there are a few assumptions here (e.g. MMU is active for example). So if you have a good reason for having these methods on PAddr we could add them...

woppopo commented:
Thanks for your reply.
I see. The source code for x86test is useful for understanding.
As a case of using a physical address, what about a bootloader written in Rust (with this crate)?

gz commented:
Hi, Yes that's a good point though. In 64bit it's not really possible to load a physical address because paging needs to be enabled before you enter long mode. However, for 32bit that's not necessarily true -- like you said one could be in real mode (albeit unlikely). So I'd say if you needed those methods on PAddr in for the 32bit part I wouldn't mind if you added them.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        const A       = bit!(5);
+        /// Dirty; if PS indicates whether software has written to the 2-MByte page referenced by this entry.
+        /// else ignored.
+        const D       = bit!(6);
```
AFAIK this one doesn't exist (must be zero) on 32bit?

woppopo replied:
It seems to exist when it references 4-MiB page. (Section 4.3 of Intel Developer's Manual)

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        /// Dirty; if PS indicates whether software has written to the 2-MByte page referenced by this entry.
+        /// else ignored.
+        const D       = bit!(6);
+        /// Page size; if set this entry maps a 2-MByte page; otherwise, this entry references a page directory.
```
With PSE I think this points to a 4M page?
https://en.wikipedia.org/wiki/Page_Size_Extension

woppopo replied:
Yes, that's true. I fixed the parts that these were 2MiB.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+/// Size of a large page (2 MiB)
+pub const LARGE_PAGE_SIZE: usize = 1024 * 1024 * 2;
+
+/// Size of a huge page (1 GiB)
```
Does not exist on 32bit, drop?

woppopo replied:
dropped.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+pub const BASE_PAGE_SIZE: usize = 4096;
+
+/// Size of a large page (2 MiB)
+pub const LARGE_PAGE_SIZE: usize = 1024 * 1024 * 2;
```
4 MiB? see below

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.0 & (BASE_PAGE_SIZE as u32 - 1)
+    }
+
+    /// Offset within the 2 MiB page.
```
4MiB?

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.0 & (LARGE_PAGE_SIZE as u32 - 1)
+    }
+
+    /// Offset within the 1 GiB page.
```
drop?

woppopo replied:
dropped.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_down(BASE_PAGE_SIZE as u32)
+    }
+
+    /// Return address of nearest 2 MiB page (lower or equal than self).
```
4MiB?

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_down(LARGE_PAGE_SIZE as u32)
+    }
+
+    /// Return address of nearest 1 GiB page (lower or equal than self).
```
drop?

woppopo replied:
dropped.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_up(BASE_PAGE_SIZE as u32)
+    }
+
+    /// Return address of nearest 2 MiB page (higher or equal than self).
```
4MiB?

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_up(LARGE_PAGE_SIZE as u32)
+    }
+
+    /// Return address of nearest 1 GiB page (higher or equal than self).
```
drop?

woppopo replied:
dropped.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_down(BASE_PAGE_SIZE as u32) == self
+    }
+
+    /// Is this address aligned to a 2 MiB page?
```
4MiB?

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+    }
+
+    /// Is this address aligned to a 1 GiB page?
+    pub fn is_huge_page_aligned(self) -> bool {
```
drop?

woppopo replied:
dropped.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.0 & (BASE_PAGE_SIZE as u32 - 1)
+    }
+
+    /// Offset within the 2 MiB page.
```
4MiB in comments

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.0 & (LARGE_PAGE_SIZE as u32 - 1)
+    }
+
+    /// Offset within the 1 GiB page.
```
drop

woppopo replied:
dropped.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_down(BASE_PAGE_SIZE as u32)
+    }
+
+    /// Return address of nearest 2 MiB page (lower or equal than self).
```
4MiB?

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_down(LARGE_PAGE_SIZE as u32)
+    }
+
+    /// Return address of nearest 1 GiB page (lower or equal than self).
```
drop

woppopo replied:
dropped.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_up(LARGE_PAGE_SIZE as u32)
+    }
+
+    /// Return address of nearest 1 GiB page (higher or equal than self).
```
drop

woppopo replied:
dropped.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_up(BASE_PAGE_SIZE as u32)
+    }
+
+    /// Return address of nearest 2 MiB page (higher or equal than self).
```
4MiB?

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_down(BASE_PAGE_SIZE as u32) == self
+    }
+
+    /// Is this address aligned to a 2 MiB page?
```
4MiB?

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_down(LARGE_PAGE_SIZE as u32) == self
+    }
+
+    /// Is this address aligned to a 1 GiB page?
```
drop?

woppopo replied:
dropped.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.0 & (BASE_PAGE_SIZE as u32 - 1)
+    }
+
+    /// Offset within the 2 MiB page.
```
4MiB?

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_down(BASE_PAGE_SIZE as u32)
+    }
+
+    /// Return address of nearest 2 MiB page (lower or equal than self).
```
4MiB?

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_up(BASE_PAGE_SIZE as u32)
+    }
+
+    /// Return address of nearest 2 MiB page (higher or equal than self).
```
4MiB?

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_down(BASE_PAGE_SIZE as u32) == self
+    }
+
+    /// Is this address aligned to a 2 MiB page?
```
4MiB?

woppopo replied:
fixed.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+    }
+
+    /// Offset within the 1 GiB page.
+    pub fn huge_page_offset(self) -> u32 {
```
drop?

woppopo replied:
dropped.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_down(LARGE_PAGE_SIZE as u32)
+    }
+
+    /// Return address of nearest 1 GiB page (lower or equal than self).
```
drop?

woppopo replied:
dropped.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_up(LARGE_PAGE_SIZE as u32)
+    }
+
+    /// Return address of nearest 1 GiB page (higher or equal than self).
```
drop?

woppopo replied:
dropped.

gz commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1527 @@
+        self.align_down(LARGE_PAGE_SIZE as u32) == self
+    }
+
+    /// Is this address aligned to a 1 GiB page?
```
drop?

woppopo replied:
dropped.

gz reviewed (commented):
Let me know if these comments make sense

woppopo commented:
Thanks for your review.
- 2 MiB page -> 4 MiB page
- Removed huge page
- Updated tests, comments and docs
- Added PAT bit (12) to PDFlags
- Added PAT bit (7) to PTFlags
- Added as_ptr and as_mut_ptr to PAddr (one case that a Rust program run in real mode is FreeDOS)

woppopo commented on src/bits32/paging.rs:
```diff
@@ -0,0 +1,1431 @@
+        let pt_val = pt & ADDRESS_MASK;
+        assert!(pt_val == pt.into());
+        assert!(pt % BASE_PAGE_SIZE == 0);
+        PDEntry(pt_val | flags.bits)
```
![image](https://user-images.githubusercontent.com/77098789/121347397-1b7bac80-c962-11eb-988e-47b65f80fb19.png)
I noticed that using PDEntry for 4MiB page has a problem.
If PS = 1, bit manipulation is required to follow to this specification, but PAddr can't represent 39:32 bits.

gz replied:
Agreed, I think that on 32bit systems the PAddr should probably be 64bits. e.g. ACPI might give you physical address ranges above 2^32, similarly PCI devices can be above the 4G limit. I haven't programmed for rust 32bit enough to know if you can just change PAddr to PAddr(u64) or if you need to split it into two low/high  variables, both 32bit in size?

woppopo replied:
It seems to be a feature to support PSE-36 and PSE-40.
I will drop it for now, but I might add support for PSE-* or PAE in the future if needed. Also PAddr is a transparent struct, so I think it's better to use low/high variables rather than to resize. (Or add ExtendedPAddr...)

gz commented:
lgtm thanks for all the work!

gz commented:
release 0.40 with these changes
